tools/code_analyzer/op_dependency.cpp

```diff
@@ -0,0 +1,396 @@
+
+  OpDependency() : ModulePass(ID) {}
+  ~OpDependency() = default;
+
+  bool runOnModule(Module& M) override {
+    // Key nodes are the nodes we keep in the result graph. They are global
+    // strings that match function schema pattern. Function names that match
+    // DebugFilters are also counted as key nodes.
+    SET keyNodes;
+
+    // Find global strings matching function schema pattern. Then use fuzz
+    // matcher to find functions related to these function schema strings.
+    GRAPH schemaStrToFunctions, functionsToSchemaStrs;
+    scanGlobalsForFunctionSchema(
+        M, &keyNodes, &schemaStrToFunctions, &functionsToSchemaStrs);
+
+    // Find all function names that match DebugFilters.
+    if (!DebugFilters.empty()) {
+      scanFunctionsForDebugFilters(M, &keyNodes);
+    }
+
+    // Extended llvm::CallGraph analysis - all references are counted.
+    GRAPH functionToFunctions;
+    scanReferenceGraph(M, &functionToFunctions);
+
+    // Simply dump all types of edges into one input graph.
+    GRAPH input, result;
+    mergeGraph(schemaStrToFunctions, &input);
+    mergeGraph(functionsToSchemaStrs, &input);
+    mergeGraph(functionToFunctions, &input);
+
+    // Calculate transitive closure and remove non-key nodes.
+    simplifyGraph(input, keyNodes, &result);
+
+    if (OutputFormat == OutputFormatType::Dot) {
+      printAsDot(std::cout, keyNodes, result);
+    } else if (OutputFormat == OutputFormatType::YAML) {
+      printAsYAML(std::cout, keyNodes, result);
+    }
+
+    return false;
+  }
+
+private:
+  // Searching for global string constants that look like function schema, e.g.:
+  // "aten::relu". Then search for places using the function schema string
+  // constant to 1) register op; 2) invoke op.
+  static void scanGlobalsForFunctionSchema(
+      Module& M, SET* keyNodes,
+      GRAPH* schemaStrToFunctions, GRAPH* functionsToSchemaStrs) {
+    for (GlobalVariable& GV : M.globals()) {
+      if (!GV.hasInitializer()) continue;
+      Constant* gvInit = GV.getInitializer();
+      if (gvInit->isNullValue()) continue;
+      ConstantDataArray* gvInitArray = dyn_cast<ConstantDataArray>(gvInit);
+      if (!gvInitArray || !gvInitArray->isCString()) continue;
+      std::string gvInitStr = gvInitArray->getAsCString().str();
+      if (!isFunctionSchemaString(gvInitStr)) continue;
+      std::string schemaStr = truncateFunctionSchemaString(gvInitStr);
+
+      // Track all seen schema strings. Useful for checking whether we miss any
+      // dependency in the logic below.
+      keyNodes->insert(schemaStr);
+
+      // Search for GV users.
+      for (auto gvUser : GV.users()) {
+        // Search for "i8* getelementptr inbounds" ConstantExpr.
+        ConstantExpr* CE = dyn_cast<ConstantExpr>(gvUser);
+        if (!CE) continue;
+
+        // Search for CE users.
+        for (auto ceUser : CE->users()) {
+          // Search for "std::basic_string" constructor.
```

Comment:
This kind of makes me wonder what would happen if we didn't actually construct `std::string` from our schema strings

Reply:
I commented it as "std::basic_string" but as long as it's some kinda reference (as I use "users()" API) it should follow the usage and scan the entire function scope starting from the usage instruction. I feel the assumption is relatively general.

Reply:
Ah, OK! Maybe the comment can be made clearer then.